File: basics/03-duplicate-characters/Solution.java
import java.util.*;

class Solution {
    public List<Character> findDuplicates(String s) {
        Set<Character> seen = new HashSet<>();
        Set<Character> duplicates = new HashSet<>();
        for (char c : s.toCharArray()) {
            if (seen.contains(c)) duplicates.add(c);
            else seen.add(c);
        }
        return new ArrayList<>(duplicates);
    }
}

File: basics/03-duplicate-characters/Solution.cpp
#include <string>
#include <vector>
#include <unordered_set>

std::vector<char> findDuplicates(const std::string& s) {
    std::unordered_set<char> seen, duplicates;
    for (char c : s) {
        if (seen.count(c)) duplicates.insert(c);
        else seen.insert(c);
    }
    return std::vector<char>(duplicates.begin(), duplicates.end());
}
